### PR TITLE
Expose array-scalar binary operators and some array and reduction functions

### PR DESCRIPTION
**Pull request by mark-poscablo**
Currently, only array-array binary arithmetic operations (`+, -, *, /`) are supported. This aims to expose the array-scalar (and vice versa) operations too.

These functions are exposed too:
- Flat
- Flip
- Moddims
- MinAll
- MaxAll

Feedback welcome!

### CODE REVIEW

File: src/Wrapper/Algorithm.cs [hunk 1]
```diff
@@ -47,6 +47,91 @@ public static Complex SumAll(Array arr)
...
+                case af_dtype.u16:
+                    Internal.VERIFY(AFAlgorithm.af_min_all(out r, out _, arr._ptr));
+                    return (ushort)r;
+                case af_dtype.c32:
+                case af_dtype.c64:
+                    Internal.VERIFY(AFAlgorithm.af_min_all(out r, out i, arr._ptr));
+                    return new Complex(r, i);
+                default:
+                    throw new NotSupportedException("Data type not supported");
+            }
```
Comment (9prady9) [inline src/Wrapper/Algorithm.cs:90]: This logic can be squashed down further.
```c#
af_dtype dtype = Internal.toDType<T>();
if ( <dtype is none of the required enums> ) {
  throw new NotSupportedException("Data type not supported");
}
double r, i;
Internal.VERIFY(AFAlgorithm.af_min_all(out r, out i, arr._ptr));
if (dtype == c32 || dtype == c64) {
    return new Complex(r, i);
} else {
    return (T)r;
}
```

File: src/Wrapper/Algorithm.cs [hunk 2]
```diff
@@ -47,6 +47,91 @@ public static Complex SumAll(Array arr)
...
+                case af_dtype.u16:
+                    Internal.VERIFY(AFAlgorithm.af_max_all(out r, out _, arr._ptr));
+                    return (ushort)r;
+                case af_dtype.c32:
+                case af_dtype.c64:
+                    Internal.VERIFY(AFAlgorithm.af_max_all(out r, out i, arr._ptr));
+                    return new Complex(r, i);
+                default:
+                    throw new NotSupportedException("Data type not supported");
+            }
```
Comment (9prady9) [inline src/Wrapper/Algorithm.cs:133]: Similar note as MinAll.

File: src/Wrapper/Array.cs [hunk 3]
```diff
@@ -163,6 +164,881 @@ public Type ElemType
 		[MethodImpl(MethodImplOptions.AggressiveInlining)]
 		public static Array operator ^(Array lhs, Array rhs) { IntPtr ptr; Internal.VERIFY(AFArith.af_bitxor(out ptr, lhs._ptr, rhs._ptr, false)); return new Array(ptr); }
 #endif
+
+#if _
```
Comment (9prady9) [inline src/Wrapper/Array.cs:168]: I have only one question in this file's change. I can understand what the logic under the `#if`'s arms are doing, but what does this `#if _` exactly check for ?

File: src/Wrapper/Data.cs [hunk 4]
```diff
@@ -517,5 +517,35 @@ public static Array Copy(Array arr)
             return new Array(ptr);
         }
         #endregion
+
+        #region Moddims
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static Array Moddims(Array arr, params int[] dims)
```
Comment (9prady9) [inline src/Wrapper/Data.cs:523]: This is more like a curious Q, is there no wrapper dim like `dim4` at the wrapper level ?

File: src/Wrapper/Algorithm.cs (hunk 1 above)
Comment (mark-poscablo) [reply src/Wrapper/Algorithm.cs:90]: That's great, thanks for catching this.

File: src/Wrapper/Array.cs (hunk 3 above)
Comment (mark-poscablo) [reply src/Wrapper/Array.cs:168]: These directives are actually intended to be used by the F# script in AutoGenTool. Everything inside `#if _` and `#else` is the "template" for code generation, and the generated code will appear between `#else` and `#endif`

File: src/Wrapper/Data.cs (hunk 4 above)
Comment (mark-poscablo) [reply src/Wrapper/Data.cs:523]: I can't speak for the original designer of this wrapper, but I surmise that the C# Array object has rich enough features that creating a new dim4 class would be redundant. See https://docs.microsoft.com/en-us/dotnet/api/system.array?view=netcore-3.1

**Pull request merged**